`approx` representation of details failed when using a `ApproxSequenceLike` which is not `list` or `tuple`

## Description

When using a custom sequence like type then the comparison of `pytest.approx` works as expected if the assert is `True`. But in case the assertion fails, the detailed output of what went wrong is broken.

Given the following example:

```python
import pytest

class MyVec3:
    def __init__(self, x, y, z):
        self._x, self._y, self._z = x, y, z

    def __len__(self):
        return 3

    def __getitem__(self, key):
        if key == 0:
            return self._x
        if key == 1:
            return self._y
        if key == 2:
            return self._z
        raise IndexError


def test_vec_approx():
    assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=2)
    assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=0.5)
```

The first assert statement is fine, as expected.

In the second assert fails, as expected.

But I get this output:

```
============================================ test session starts ============================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.3.0
rootdir: /home/user
collected 1 item                                                                                            

test_approx.py F                                                                                      [100%]

================================================= FAILURES ==================================================
______________________________________________ test_vec_approx ______________________________________________

    def test_vec_approx():
        assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=2)
>       assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=0.5)
E       AssertionError: assert <test_approx....x7f657f208880> == approx([1 ± 5... 3 ± 5.0e-01])
E         (pytest_assertion plugin: representation of details failed: /home/user/venv/lib/python3.10/site-packages/_pytest/python_api.py:342: TypeError: object of type 'ApproxScalar' has no len().
E          Probably an object has a faulty __repr__.)

test_approx.py:22: AssertionError
========================================== short test summary info ==========================================
FAILED test_approx.py::test_vec_approx - AssertionError: assert <test_approx....x7f657f208880> == approx([1 ± 5... 3 ± 5.0e-01])
============================================= 1 failed in 0.01s =============================================
```

## Setup
```
Package        Version
-------------- -------
exceptiongroup 1.2.0
iniconfig      2.0.0
packaging      23.2
pip            22.0.2
pluggy         1.3.0
pytest         7.4.4
setuptools     59.6.0
tomli          2.0.1
```

Also happens with pytest version 7.1.2 and 7.1.3

OS is Ubuntu 22.04

Replacing `_recursive_sequence_map` with

```python
def _recursive_sequence_map(f, x):
    """Recursively map a function over a sequence of arbitrary depth"""
    if (
        hasattr(x, "__getitem__")
        and isinstance(x, Sized)
        # Type ignored because the error is wrong -- not unreachable.
        and not isinstance(x, STRING_TYPES)  # type: ignore[unreachable]
    ):
        if isinstance(x, (list, tuple)):
            seq_type = type(x)
        else:
            seq_type = list
        return seq_type(_recursive_sequence_map(f, xi) for xi in x)
    else:
        return f(x)
```

would result in

```
    def test_vec_approx():
        assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=2)
>       assert MyVec3(0, 1, 2) == pytest.approx(MyVec3(1, 2, 3), abs=0.5)
E       assert <test_approx....x7f9c09e6ca60> == approx([1 ± 5... 3 ± 5.0e-01])
E         comparison failed. Mismatched elements: 3 / 3:
E         Max absolute difference: 1
E         Max relative difference: inf
E         Index | Obtained | Expected   
E         0     | 0        | 1 ± 5.0e-01
E         1     | 1        | 2 ± 5.0e-01
E         2     | 2        | 3 ± 5.0e-01
```

```
self = approx([1 ± 5.0e-01, 2 ± 5.0e-01, 3 ± 5.0e-01])
other_side = <approx.TestApprox.test_strange_sequence.<locals>.MyVec3 object at 0x7f8237a32d50>

    def _repr_compare(self, other_side: Sequence[float]) -> List[str]:
        import math
    
        if len(self.expected) != len(other_side):
            return [
                "Impossible to compare lists with different sizes.",
                f"Lengths: {len(self.expected)} and {len(other_side)}",
            ]
    
        approx_side_as_map = _recursive_sequence_map(self._approx_scalar, self.expected)
    
>       number_of_elements = len(approx_side_as_map)
E       TypeError: object of type 'ApproxScalar' has no len()

../../src/_pytest/python_api.py:332: TypeError
```

i have a error indicating reproducer i'll extract a is_sequencelike and use that

--- a/changelog/11797.bugfix.rst
+++ b/changelog/11797.bugfix.rst
@@ -1,0 +1,1 @@
+:func:`pytest.approx` now correctly handles :class:`Sequence <collections.abc.Sequence>`-like objects.

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -129,6 +129,8 @@ def _recursive_sequence_map(f, x):
     if isinstance(x, (list, tuple)):
         seq_type = type(x)
         return seq_type(_recursive_sequence_map(f, xi) for xi in x)
+    elif _is_sequence_like(x):
+        return [_recursive_sequence_map(f, xi) for xi in x]
     else:
         return f(x)
 
@@ -721,11 +723,7 @@ def approx(expected, rel=None, abs=None, nan_ok: bool = False) -> ApproxBase:
     elif _is_numpy_array(expected):
         expected = _as_numpy_array(expected)
         cls = ApproxNumpy
-    elif (
-        hasattr(expected, "__getitem__")
-        and isinstance(expected, Sized)
-        and not isinstance(expected, (str, bytes))
-    ):
+    elif _is_sequence_like(expected):
         cls = ApproxSequenceLike
     elif isinstance(expected, Collection) and not isinstance(expected, (str, bytes)):
         msg = f"pytest.approx() only supports ordered sequences, but got: {expected!r}"
@@ -734,6 +732,14 @@ def approx(expected, rel=None, abs=None, nan_ok: bool = False) -> ApproxBase:
         cls = ApproxScalar
 
     return cls(expected, rel, abs, nan_ok)
+
+
+def _is_sequence_like(expected: object) -> bool:
+    return (
+        hasattr(expected, "__getitem__")
+        and isinstance(expected, Sized)
+        and not isinstance(expected, (str, bytes))
+    )
 
 
 def _is_numpy_array(obj: object) -> bool:

--- a/testing/python/approx.py
+++ b/testing/python/approx.py
@@ -954,6 +954,43 @@ class TestApprox:
         with pytest.raises(TypeError, match="only supports ordered sequences"):
             assert {1, 2, 3} == approx({1, 2, 3})
 
+    def test_strange_sequence(self):
+        """https://github.com/pytest-dev/pytest/issues/11797"""
+        a = MyVec3(1, 2, 3)
+        b = MyVec3(0, 1, 2)
+
+        # this would trigger the error inside the test
+        pytest.approx(a, abs=0.5)._repr_compare(b)
+
+        assert b == pytest.approx(a, abs=2)
+        assert b != pytest.approx(a, abs=0.5)
+
+
+class MyVec3:  # incomplete
+    """sequence like"""
+
+    _x: int
+    _y: int
+    _z: int
+
+    def __init__(self, x: int, y: int, z: int):
+        self._x, self._y, self._z = x, y, z
+
+    def __repr__(self) -> str:
+        return f"<MyVec3 {self._x} {self._y} {self._z}>"
+
+    def __len__(self) -> int:
+        return 3
+
+    def __getitem__(self, key: int) -> int:
+        if key == 0:
+            return self._x
+        if key == 1:
+            return self._y
+        if key == 2:
+            return self._z
+        raise IndexError(key)
+
 
 class TestRecursiveSequenceMap:
     def test_map_over_scalar(self):
@@ -981,3 +1018,6 @@ class TestRecursiveSequenceMap:
             (5, 8),
             [(7)],
         ]
+
+    def test_map_over_sequence_like(self):
+        assert _recursive_sequence_map(int, MyVec3(1, 2, 3)) == [1, 2, 3]

